Initial Airbnb stylelint shared config `stylelint-config-airbnb`

ntwb opened the pull request:
See #13

ntwb commented:
A couple of notes:

1. I structured the folders in the same way as https://github.com/airbnb/javascript, this should facilitatte publishing the `styleline-config-airbnb` seperately from this repo, as is done for `eslint-config-airbnb`

2. Sorry no tests, I had to remove them as NPM was exploding and not installing packages, this can be easily revisited later, most things here in this PR are a cut and paste from https://github.com/ntwb/stylelint-config-wordpress with the tests pulled, and the config edited

DanielaValero commented:
Hello guys,
Does any of you have time to check this pull request out? 

Having an official config for the airbnb css style guide would be very helpful for the community.
We would get the benefits of it, and also would be able to make pull requests to enhance it.


Thanks in advance

ljharb commented on packages/stylelint-config-airbnb/.eslintrc:
```diff
@@ -0,0 +1,3 @@
+{
+  "extends": "airbnb"
+}
```
Could you make sure all these files have trailing newlines?

ntwb replied:
Done in https://github.com/airbnb/css/pull/23/commits/ff18d84c4b306ecb6db23d3957337c16b5b9591b

ljharb commented on packages/stylelint-config-airbnb/package.json:
```diff
@@ -0,0 +1,38 @@
+    "eslint-plugin-import": "^1.8.1",
+    "eslint-plugin-jsx-a11y": "^1.2.3",
+    "eslint-plugin-react": "^5.1.1",
+    "stylelint": "^6.5.1",
```
should stylelint and stylelint-scss also be a peerDependency?

ntwb replied:
No, this shouldn't be needed, a few of the other stylelint configurations required the peerDependency to work around an Atom Editor issue.

ljharb replied:
Right - what I mean is, this config is likely compatible with specific versions of stylelint - making it a peer dep allows us to convey "which versions we're compatible with" in a way that fits easily with npm.

If a package wouldn't ever reasonably be installed without X, and it requires X to work, then X must be a dep or a peer dep :-)

ljharb commented:
@ntwb Sorry for the delay on this!

Would you mind addressing these comments, and updating all the deps to the latest versions?

In addition, adding documentation as suggested by @DanielaValero would be quite helpful.

ljharb commented on packages/stylelint-config-airbnb/.eslintrc:
```diff
@@ -0,0 +1,3 @@
+{
+  "extends": "airbnb"
```
let's add `"root": true` to this as well

ntwb commented:
> Would you mind addressing these comments, and updating all the deps to the latest versions?

@ljharb I've bumped them all apart from ESLint to 3.x per https://github.com/airbnb/javascript/issues/936

ljharb commented on packages/stylelint-config-airbnb/package.json:
```diff
@@ -0,0 +1,38 @@
+  "homepage": "https://github.com/airbnb/css#readme",
+  "devDependencies": {
+    "eslint": "^2.13.1",
+    "eslint-config-airbnb": "^9.0.1",
```
actually since we don't use react in this project, you could use `eslint-config-airbnb-base` - the latest of which does support eslint v3 :-)

ntwb commented:
@ljharb Is that everything?

I've got more enhancements that can be added such as tests and some more SCSS rules but I'd rather look at that over the weekend and send those via PR's...

ljharb commented on packages/stylelint-config-airbnb/package.json:
```diff
@@ -26,8 +26,8 @@
+    "eslint-config-airbnb-base": "^4.0.0",
     "eslint-plugin-import": "^1.10.2",
     "eslint-plugin-jsx-a11y": "^1.5.5",
     "eslint-plugin-react": "^5.2.2",
```
this means the react and jsx-a11y plugins aren't needed anymore

ljharb commented:
I don't think this will be able to be merged before next week - I need to run it by some colleagues first. It's totally fine if you want to add those things to a separate PR (or to this one if it's not merged by the time you're done).

ntwb commented:
Cool, I'll just keep pushing commits here then 👍 

Not sure if you can turn Travis CI on for this repo but now that there's a `.travis.yml` file I'll add some tests also.

ljharb commented:
Good call - it's now enabled whenever there's a .travis.yml at the root. Please continue to mirror the "javascript" repo's organization :-)

ntwb commented:
> Good call - it's now enabled whenever there's a .travis.yml at the root. 

At the root won't work here then will it?

> Please continue to mirror the "javascript" repo's organization :-)

👍

ljharb commented:
@ntwb The JS repo has two subpackages and a root .travis.yml file - you can probably just copy + modify the one from there.

DanielaValero commented:
Hello @ntwb 

I cloned your fork and made two commits to apply two fixes into the code, however the pull request can be made only to the original repo. Here the two relevant commits:

https://github.com/DanielaValero/css/commits/master


Also I've set up a stylelint config with all the config of airbnb, plus some best practices:

https://gist.github.com/DanielaValero/b3a72b2f3d0c85c1985d5d2d7497f347


I was trying to come up with a regex to lint the naming convention for the variables, but did not figure it out quite right.. This is the last state I could get to:

'scss/dollar-variable-pattern': '\b[a-z]+(?:-)+(\b[a-z]+(?:-))*'

ntwb commented:
@DanielaValero Thanks, I've merged those two commits  👍 

I messed up a pull request, but @DanielaValero you should be able to create pull requests against my fork and they will be added here to this pull request once merged, I think 😏 

Do you want to try that with some of the changes you have in your gist?

ljharb commented on packages/stylelint-config-airbnb/index.js:
```diff
@@ -0,0 +1,29 @@
+    'stylelint-scss',
+  ],
+  rules: {
+    'at-rule-empty-line-before': 'never',
```
i'm kind of surprised that this plugin doesn't have a prefix on its rule names - ie, `stylelint-scss/indentation` - to avoid the risk of colliding with core eslint rules :-/

ntwb replied:
It's 3am and much wine so forgive me if I interpret this a little wrong 😏 

* The rule `at-rule-empty-line-before` your comment refers to is a core stylelint rule, it is **not** a rule from the `stylelint-scss` plugin, you can see a prefixed example of this rule for that plugin [here](https://github.com/airbnb/css/pull/23/files/e7956634495732478f2e76ada527e1ae80b6c997#diff-c8daf3b8e5463995761e914cf32b47a5R22)

ljharb replied:
right but it's still appearing in an eslint config, so it could still conflict with a core *eslint* rule. i'm saying core styleint rules should be prefixed with "stylelint/"

ntwb replied:
I'll reiterate my previous comment, it's late, its 3am and much wine has passed 😬 

This is **not** an ESLint configuration file, this is a stylelint configuration file

This repo has nothing to do with ESLint (aside ESLint lints the codebase)

ljharb replied:
ok that's just my confusion then :-) in that case, eslint shouldn't be a peer dependency.

ntwb replied:
Hehehehe

ljharb commented on packages/stylelint-config-airbnb/package.json:
```diff
@@ -0,0 +1,41 @@
+    "eslint": "^3.0.1",
+    "eslint-config-airbnb-base": "^4.0.0",
+    "eslint-plugin-import": "^1.10.2",
+    "eslint-plugin-jsx-a11y": "^1.5.5",
```
this dep isn't needed when using the base config

ntwb replied:
Done in https://github.com/ntwb/css/commit/aa6bb964219dfb053f86d95777045687fda08b80

ljharb commented on packages/stylelint-config-airbnb/package.json:
```diff
@@ -0,0 +1,41 @@
+    "eslint-config-airbnb-base": "^4.0.0",
+    "eslint-plugin-import": "^1.10.2",
+    "eslint-plugin-jsx-a11y": "^1.5.5",
+    "eslint-plugin-react": "^5.2.2",
```
this dep isn't needed when using the base config

ntwb replied:
Done in https://github.com/ntwb/css/commit/aa6bb964219dfb053f86d95777045687fda08b80

ljharb commented on packages/stylelint-config-airbnb/package.json:
```diff
@@ -0,0 +1,41 @@
+    "stylelint-scss": "^1.2.1"
+  },
+  "peerDependencies": {
+    "stylelint": "^6.9.0",
```
we should also probably add eslint as a peer dep

ntwb replied:
Done in https://github.com/ntwb/css/commit/43f6dbfdaa3a906df239be6a9d951938d7e8e57d

ntwb replied:
Reverted per https://github.com/airbnb/css/pull/23/files#r70673569

DanielaValero commented:
@ntwb sounds like a plan, I'll do a pull request on your fork

ljharb commented on packages/stylelint-config-airbnb/.travis.yml:
```diff
@@ -0,0 +1,11 @@
+  directories:
+    - node_modules
+node_js:
+  - '6'
```
let's add `5` and `7` for completeness

ntwb replied:
Done in https://github.com/airbnb/css/pull/23/commits/ea3ef3d0edda84a638633ee079941a63d4b192b4

lencioni commented:
Unfortunately, there are a lot of scss-lint rules that are enabled by default, so they are implicitly part of our scss-lint configuration. It would be nice if the stylelint configuration captured these. You can see the default set here: https://github.com/brigade/scss-lint/blob/master/config/default.yml

ntwb commented:
@lencioni Not entirely sure what you mean here, there are only two SCSS lint rules in the PR currently, `scss/dollar-variable-pattern` and `scss/at-extend-no-missing-placeholder`

Any reasons as to what is what is not in this initial pull request is because there is no other documentation of the Airbnb CSS elsewhere... 

I'd also like not have to continue to iterate on this one single pull request, it would be nice to get this merged initially and then build on it with further pull requests.

lencioni commented:
I may be misunderstanding the intent here, but if it is to produce a stylelint configuration file that would be equivalent the scss-lint configuration in this repo, then I am suggesting that it capture the spirit of all the rules that scss-lint enables by default in addition to the rules that are explicitly enabled in the scss-lint configuration in this repo. They don't need to be the same rules, but they should cover the same scenarios.

ntwb commented:
@lencioni Cool, I'm with you now 👍 

There will be a bit of time involved in going through that set of default scss-lint rules to their equivalent stylelint-css and stylelint-scss rules. I'll find a place/repo to document these so that I need only do this once and is also then available for others as a reference 💥

lencioni commented:
If you are going through the trouble anyway, it might not be a bad idea to pull all of the default rules into the scss-lint config here as well.

ntwb commented:
Yeah I will, it can be either in this PR, or a follow up PR, I add to add my last commit via the GutHub GUI as my local branch is full of conflicts 😖 

Iteration is our friend ;)

anna944 commented:
**spielen! http://cutt.us/1001spiele**

fsmaia commented:
Any help wanted? Would be very nice to see `stylelint-config-airbnb` published in NPM.

TrevorSayre commented:
@ntwb What's the reasoning behind the blacklist for border?

```
    "declaration-property-value-blacklist": {
      "/^border/": ["none"]
    }
```

ntwb commented:
To disallow the value `none` for the `border` properties. 

There was more than likely some Airbnb code I reference that showed the preference was `0` rather than `none`

TrevorSayre commented:
@ntwb I understood what the code does, was curious why you wanted to prevent `none` for `border`. However, some [quick research](https://codepen.io/denilsonsa/pen/LkdHh/) shows that `0` is nearly always preferable save for when dealing with tables where `hidden` may be preferred.

ntwb commented:
Thanks @TrevorSayre what I meant was I must have made that decision based off of an Airbnb code example somewhere, it also could have been a decision I made as we'd discussed this rule and the `none` vs `0` debate at @stylelint considerably, I honestly I cannot remember, it was so long ago 😏

TrevorSayre commented:
@ljharb What's stalling this PR? Is there anything that still needs to be done to achieve acceptance?

ljharb commented:
@TrevorSayre My comment [here](https://github.com/airbnb/css/pull/23#pullrequestreview-11480177) describes what's still needed.

ntwb commented:
The Travis CI changes you requested there @ljharb were added the same day were they not?

ljharb commented:
No, i requested "at the root" (of the repo) since travis doesn't know about "packages" - similarly to how the javascript repo does it.

ntwb commented:
I've got a heap of merge conflicts, can you just merge this pull request so it can be iterated on please

ljharb commented:
@ntwb then can you check the "allow edits" box on the right hand column, so i can rebase it for you?

ntwb commented:
> then can you check the "allow edits" box on the right hand column, so i can rebase it for you?

Done, I never knew that was a thing, it's an awesome thing matter of fact 👍

ljharb commented:
k, rebased. travis.yml still needs to be fixed.

cmalard commented:
@ntwb @ljharb Hello there, is there still something to do before merge ? :-)
Would be very nice to see `stylelint-config-airbnb` published in NPM :beers:

ljharb commented:
@cmalard https://github.com/airbnb/css/pull/23#issuecomment-306679267

chlorophyllkid commented:
Hey, great work so far!
Is there is a chance this still get's done?

I assume the travis.yml has to be changed to something like the [airbnb/javascript/.travis.yml](https://github.com/airbnb/javascript/blob/master/.travis.yml), right?

ntwb commented:
Closing in favour of #56
